### PR TITLE
NOVER-5: add authz http mw

### PR DESCRIPTION
**Pull request by cbarik-infoblox**
The `http_opa` package contains the new http middleware implementation for authorising http requests.
Most of the public methods/functions/interfaces have been copied from the `grpc_opa` package to allow for backward compatibility.
Some of the util functions have been moved to `common` folder to be reused by grpc and http and yet a stub file has been created to avoid build issues for the existing consumers.

For the opa client, it expects a jwt token in the `Authorization` header, prefixed with `Bearer ` and an optional `X-Request-ID` header. 

Unauthorised requests will usually be rejected with a status of `401 Unauthorized`

### CODE REVIEW

File: common/claim/claim.go [hunk 1]
```diff
@@ -20,9 +20,6 @@ func NullClaimsVerifier([]string, []string) (string, []error) {
 func UnverifiedClaimFromBearers(bearer, newBearer []string) (string, []error) {
 	_, validBearer, bearerErrorList := atlas_claims.ParseUnverifiedClaimsFromJwtStringsRaw(bearer)
 	_, validNewBearer, newBearerErrorList := atlas_claims.ParseUnverifiedClaimsFromJwtStringsRaw(newBearer)
-	if len(newBearerErrorList) > 0 || len(bearerErrorList) > 0 {
-		//fishy Should not have multiple newBearers
-	}
```
Comment (kumaya) [inline common/claim/claim.go:25]: Add a `TODO` for checks. Do not remove.

File: go.mod [hunk 2]
```diff
@@ -9,11 +9,19 @@ require (
...
 	go.opencensus.io v0.23.0
 	golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd
 	google.golang.org/grpc v1.44.0
 )
 
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/stretchr/objx v0.1.1 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
```
Comment (kumaya) [inline go.mod:22]: are these new packages required for http middleware ?

Comment (cbarik-infoblox) [reply go.mod:22]: [out.txt](https://github.com/infobloxopen/atlas-authz-middleware/files/14163452/out.txt)
this is the graph file for the dependencies, as per the file, these might have been introduced due to `go.uber.org/multierr` or `github.com/stretchr/testify`. Middleware is using these libraries

File: http_opa/authorizer/authorizer.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		})
+		span.End()
+		logger.WithFields(log.Fields{
+			"opaResp": opaResp,
+			"elapsed": time.Since(now),
+		}).Debug("authorization_result")
+	}()
+	if err != nil {
+		return false, ctx, err
+	}
```
Comment (drewwells) [inline http_opa/authorizer/authorizer.go:193]: I would comment out these traces, and remove references to grpc. `http_opa` shouldn't require any version of grpc. Commenting out the traces will also remove the dependence on opentracing found in here. Maybe we should publish http_opa as a separate module too... so people aren't bound to the grpc version being imported in other parts of atlas-authz-middleware.

Comment (rchowinfoblox) [reply http_opa/authorizer/authorizer.go:193]: Are we sure we want to remove the opentracing?  Wasn't there a requirement to add them?

**Comment by rchowinfoblox** [comment]
@cbarik-infoblox you mention "Most of the public methods/functions/interfaces have been copied".  I'm not what you mean by this.  My concern if there is ANY public method/function/interface/type/variable/constant/whatever that has been removed, it may break backwards compatibility.

**Comment by cbarik-infoblox** [comment]
> @cbarik-infoblox you mention "Most of the public methods/functions/interfaces have been copied". I'm not what you mean by this. My concern if there is ANY public method/function/interface/type/variable/constant/whatever that has been removed, it may break backwards compatibility.

@rchowinfoblox when I said copied, I meant that the new package `http_opa` is created by taking most of the functionalities from the existing package `grpc_opa` while keeping the latter intact. Whatever has been moved out of `grpc_opa` is compensated by creating the stub inside the same package. Hope this clarifies

File: http_opa/authorizer/authorizer.go (hunk 3 above)
Comment (kumaya) [reply http_opa/authorizer/authorizer.go:193]: @cbarik-infoblox Please remove grpc related dependencies from http middleware. While we would need tracing for http; add a TODO and work towards getting the initial functionality.

**Comment by kumaya** [comment]
> My concern if there is ANY public method/function/interface/type/variable/constant/whatever that has been removed, it may break backwards compatibility

What is your suggestion, @rchowinfoblox ? Shouldn't the current test cases be capable of detecting those issues?

**Comment by rchowinfoblox** [comment]
> What is your suggestion, @rchowinfoblox ? Shouldn't the current test cases be capable of detecting those issues?

I'm not sure our current test cases provide 100% coverage.  Anyways I'm good with @cbarik-infoblox 's reply above.

File: go.mod (hunk 2 above)
Comment (drewwells) [reply go.mod]: this is middleware, lets use standard go libraries.

Comment (cbarik-infoblox) [reply go.mod]: Hi @drewwells , I see that these two libraries are widely used in infoblox.
https://github.com/search?q=org%3AInfoblox-CTO%20github.com%2Fstretchr%2Ftestify&type=code

https://github.com/search?q=org%3AInfoblox-CTO%20go.uber.org%2Fmultierr&type=code

**Pull request closed without merging**